contract_comm/currency/currency.go

```diff
@@ -94,97 +94,127 @@ const (
+		return nil, errors.ErrExchangeRateZero
+	}
+	return &ExchangeRate{numerator, denominator}, nil
+}
 
-	if *currencyFrom == celoGoldAddress {
-		// This function shouldn't really be called with the token's address, but if it is the value
-		// is correct, so return nil as the error
-		return val, nil
-	} else if err != nil {
-		log.Error(err.Error())
-		return val, err
+// ToBase converts from token to base
+func (er *ExchangeRate) ToBase(tokenAmount *big.Int) *big.Int {
+	// check noop rate (cGLD rate)
+	if er == nil {
+		return new(big.Int).Set(tokenAmount)
 	}
-	return Convert(val, currencyFrom, nil)
+
+	return new(big.Int).Div(new(big.Int).Mul(tokenAmount, er.denominator), er.numerator)
 }
 
-// NOTE (jarmg 4/24/19): values are rounded down which can cause
-// an estimate to be off by 1 (at most)
-func Convert(val *big.Int, currencyFrom *common.Address, currencyTo *common.Address) (*big.Int, error) {
-	exchangeRateFrom, err1 := getExchangeRate(currencyFrom)
-	exchangeRateTo, err2 := getExchangeRate(currencyTo)
+// FromGold converts from base to token
+func (er *ExchangeRate) FromBase(goldAmount *big.Int) *big.Int {
+	// check noop rate (cGLD rate)
+	if er == nil {
+		return new(big.Int).Set(goldAmount)
+	}
 
-	if err1 != nil || err2 != nil {
-		log.Error("Convert - Error in retreiving currency exchange rates")
-		if err1 != nil {
-			return nil, err1
-		}
-		if err2 != nil {
-			return nil, err2
-		}
+	return new(big.Int).Div(new(big.Int).Mul(goldAmount, er.numerator), er.denominator)
+}
+
+func (er *ExchangeRate) CmpValues(amount *big.Int, anotherTokenAmount *big.Int, anotherTokenRate *ExchangeRate) int {
+	// if both rates are noop rate (cGLD rate), compare values
+	if er == nil && anotherTokenRate == nil {
+		return amount.Cmp(anotherTokenAmount)
 	}
 
-	// Given value of val and rates n1/d1 and n2/d2 the function below does
-	// (val * d1 * n2) / (n1 * d2)
-	numerator := new(big.Int).Mul(val, new(big.Int).Mul(exchangeRateFrom.Denominator, exchangeRateTo.Numerator))
-	denominator := new(big.Int).Mul(exchangeRateFrom.Numerator, exchangeRateTo.Denominator)
+	// if both exchangeRate are the same
+	if er != nil && anotherTokenRate != nil && *er == *anotherTokenRate {
+		return amount.Cmp(anotherTokenAmount)
+	}
 
-	if denominator.Sign() == 0 {
-		log.Error("Convert - Error in retreiving currency exchange rates")
-		return nil, errors.ErrExchangeRateZero
+	// Below code block is basically evaluating this comparison:
+	// amount * er.denominator / er.numerator < anotherTokenAmount * anotherTokenRate.denominator / anotherTokenRate.numerator
+	// It will transform that comparison to this, to remove having to deal with fractional values.
+	// amount * er.denominator * anotherTokenRate.numerator < anotherTokenAmount * anotherTokenRate.denominator * er.numerator
+
+	// when either er or anotherTokenRate are nil, we assume rate = 1/1 (as in cGLD)
+	var leftSide, rightSide *big.Int
+	if er == nil {
```

Comment:
The need for 3 cases here makes me miss the {1, 1} exchange rate

Reply:
One way to make this section simpler is to add an `ExchangeRate.Numerator()` function that returns `common.Big1` if `er == nil`. (I mentioned this in another comment as well)

Reply:
removed `nil` as a valid value for exchange rate